colossus/src/main/scala/colossus/controller/Pipe.scala

```diff
@@ -0,0 +1,411 @@
+package colossus
+package controller
+
+import core.DataBuffer
+import akka.util.ByteString
+import java.util.LinkedList
+import scala.util.{Try, Success, Failure}
+
+import service.{Callback, UnmappedCallback}
+
+
+trait Transport {
+  //can be triggered by either a source or sink, this will immediately cause
+  //all subsequent pull and pushes to return an Error
+  def terminate(reason: Throwable)
+
+  def terminated : Boolean
+
+  def isClosed: Boolean
+}
+
+/**
+ * A Sink is the write side of a pipe.  It allows you to push items to it,
+ * and will return whether or not it can accept more data.  In the case where
+ * the pipe is full, the Sink will return a mutable Trigger and you can
+ * attach a callback to for when the pipe can receive more items
+ */
+trait Sink[T] extends Transport {
+  def push(item: T): PushResult
+
+  def isFull: Boolean
+
+  //after this is called, data can no longer be written, but can still be read until EOS
+  def complete(): Try[Unit]
+
+  def feed(iterator: Iterator[T]) {
+    var continue = true
+    while (iterator.hasNext && continue) {
+      this.push(iterator.next) match {
+        case PushResult.Filled(sig) => {
+          continue = false
+          sig.react { feed(iterator) }
+        }
+        case other: PushResult.NotPushed => {
+          //the pipe is closed or dead, so just stop feeding
+          continue = false
+        }
+        case other => {}
+      }
+    }
+    if (!iterator.hasNext) {
+      this.complete()
+    }
+  }
+
+    
+
+}
+
+/**
+ * A Source is the read side of a pipe.  You provide a handler for when an item
+ * is ready and the Source will call it.  Note that if the underlying pipe has
+ * multiple items ready, onReady will only be called once.  This is so that the
+ * consumer of the sink can implicitly apply backpressure by only pulling when
+ * it is able to
+ */
+trait Source[T] extends Transport {
+  def pull(onReady: Try[Option[T]] => Unit)
```

Comment:
I'm going to leave this as-is for now since I already have another branch that changes the logic of `Pull` so it returns an ADT similar to `push`